### PR TITLE
tvu and tpu timeout on joining its microservices

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

During validator shutdown, the validator could hang and deadlock on joining microservices inside TVU and TPU. During my test, this deadlock happens approximately about 1 out 10 runs.

This is because of the blocking recev calls on the signals receiver in those microservices, one example is here.
https://github.com/solana-labs/solana/blob/eb65ffb7798e26dc559e31f480ab20e297e951ac/core/src/shred_fetch_stage.rs#L87
There could be other places. I will try to address them in future.

Here, in this pull request, I added a timeout of 10s for joining TVU and TPU in case that they are in deadlock when shutting down. I have run the coverage test suite locally last night for `1173` times without hanging. 

#### Summary of Changes

Fixes #

### CODE REVIEW

**Comment by BGZHU1** [comment]
Hello @HaoranYi, I have some issue with this test:
I get the following message:
test validator::tests::validator_exit has been running for over 60 seconds
# Received cancellation signal, interrupting

Here is the test link failed:
https://buildkite.com/solana-labs/solana/builds/69580#7b9480ae-54f2-4498-b6be-ccbb7aaf2c0a

**Comment by t-nelson** [comment]
this isn't solving anything right, just failing faster?

**Comment by HaoranYi** [comment]
> this isn't solving anything right, just failing faster?

yup. you are correct. still trying to reproduce it..

File: core/src/tpu.rs
Comment (t-nelson) [inline core/src/tpu.rs:42]:
```suggestion
const TPU_THREADS_JOIN_TIMEOUT_SECONDS: u64 = 10;
```

File: core/src/tpu.rs
Comment (t-nelson) [inline core/src/tpu.rs:222]:
```suggestion
        // exit can deadlock. put an upper-bound on how long we wait for it
```

File: core/src/tvu.rs
Comment (t-nelson) [inline core/src/tvu.rs:70]:
```suggestion
const TVU_THREADS_JOIN_TIMEOUT_SECONDS: u64 = 10;
```

File: core/src/tvu.rs
Comment (t-nelson) [inline core/src/tvu.rs:375]:
```suggestion
        // exit can deadlock. put an upper-bound on how long we wait for it
```

File: core/src/validator.rs
Comment (t-nelson) [inline core/src/validator.rs:1879]:
```suggestion
        // exit can deadlock. put an upper-bound on how long we wait for it
```

File: core/src/tpu.rs
Comment (t-nelson) [inline core/src/tpu.rs:225]: should we just panic here too?  may as well be explicit  (same with the TPU instance)

Comment (HaoranYi) [reply core/src/tpu.rs:225]: Good question. I had thought about this too and decided to not panic. 

1. when deadlock, panic will still make the test fail, unless we mark panic expected in the test. If we mark panic expected in test, which will mask the actual panic in the test, and prevent us from detecting other deadlock that may have introduced in future. 
2. panic may prevent other services clean code to run. 
3. since we are already shutting down, printing an error message could be enough.

**Review by t-nelson — APPROVED** [review]
thanks! r+ ci

**Comment by behzadnouri** [comment]
@HaoranYi are these meant to be a permanent solution?
Or is there any tracking issue to address the root cause?
https://github.com/solana-labs/solana/blob/53695ecd21d19d16268a3cba58f7681436cf1de3/core/src/tpu.rs#L261-L266

**Comment by HaoranYi** [comment]
No. It is a stopgap. Yes, we can create a issue to track down the root
cause of the hang.

**Comment by HaoranYi** [comment]
https://github.com/solana-labs/solana/issues/25933
This is the tracking issue.

**Comment by behzadnouri** [comment]
This is causing segfault in rocksdb when running tests: https://github.com/solana-labs/solana/issues/25941#issuecomment-1162480391
Can we please revert this change and instead address the root cause?

**Comment by HaoranYi** [comment]
OK. You can revert it.

**Comment by behzadnouri** [comment]
@HaoranYi I would appreciate it if going forward more time and effort is spent into addressing root causes as opposed to adding band-aids or taking shortcuts.
Several engineers spent time chasing mysterious segfaults in https://github.com/solana-labs/solana/issues/25941
I myself have been blocked on #25807 since it was opened (more than 2 weeks) because it was hitting that segfault scenario very often.

I personally would have much preferred spending that time correctly addressing timeouts in `validator_exit` test as opposed to being blocked on investigating segfaults on a very unrelated change.

**Comment by HaoranYi** [comment]
@behzadnouri 
I disagree with your remarks. I don't think this change is the culprit of your segfaults. And I still don't understand how revert this change will fix your problem. Do you actually see timeout log in your tests when you have segmentation fault with this change?

**Comment by behzadnouri** [comment]
> Do you actually see timeout log in the tests when you have seen segmentation fault with this change?

Yes, please also see this comment: https://github.com/solana-labs/solana/issues/25941#issuecomment-1162480391

Even without the segfault issue that transpired, it is obvious that this commit was hacking some band-aids instead of addressing the root cause. Going forward I would appreciate it if when you can't or don't want to debug and correctly patch some issues, rather than commits like this, assign the issue to either myself or someone else on the team.

As I said in the previous comment, I would have much preferred that I had spent the past 2 weeks time debugging timeouts in `validator_exit` test as opposed to being blocked on investigating segfaults on a very unrelated change which was very hard to track down where the culprit is.

**Comment by HaoranYi** [comment]
@behzadnouri 
OK. I assigned https://github.com/solana-labs/solana/issues/25933 to you. Worth another pair of eyes to look.

**Comment by HaoranYi** [comment]
@behzadnouri 
Because the segfault occurs after 10s forcing shutdown, it seems that the hang might be in the shredder related stages in tpu/tvu. Since you are familiar with this piece of code, it is better for you to take a  look. Perhaps, you can spot the bug there. If you don't find anything, I can help to take a look next week.

**Pull request merged**
